Kiosk SDK V2 Docs (#13571)
MystenLabs/sui: sdk/docs/pages/kiosk/kiosk-client/kiosk-transaction/managing.mdx

# Managing Owned Kiosk

> `KioskClient` helps in making managing a kiosk simple.

Reviewer: ```suggestion
`KioskClient` helps in managing a kiosk.
```

> You need to follow the steps explained in the [Kiosk Transaction section](kiosk-transaction) to

Reviewer: ```suggestion
You need to follow the steps explained in the [Kiosk Transaction section](#kiosk-transaction) to
```

create a `KioskTransaction`.

## Available functions

### take

> Removes an item from the Kiosk and returns a `TransactionArgument` to use it in a different PTB

Reviewer: ```suggestion
Removes an item from the Kiosk and returns a `TransactionArgument` to use it in a different Programmable Transaction Block (PTB)
```

call.

```typescript
const item = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

/// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

> // take item from kiosk.

Reviewer: ```suggestion
// Take item from kiosk.
```

const item = kioskTx.take({
	itemId: item,
	itemType,
});

> // do something with `item`, e.g. transfer it to someone else.

Reviewer: ```suggestion
// Do something with `item`, like transfer it to someone else.
```

txb.transferObjects([item], txb.pure('address_to_transfer_the_object'));

> // finalize the kiosk Tx.

Reviewer: ```suggestion
// Finalize the kiosk Tx.
```

kioskTx.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```

await signAndExecuteTransactionBlock({ tx: txb });
```

### transfer

Similar to `take`, but transfers the item to an address internally.

```typescript
const item = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

> /// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.

Reviewer: ```suggestion
/// Assume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
```

const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

// take item from kiosk.
kioskTx
	.transfer({
		itemId: item,
		itemType,
		address: 'address_to_transfer_the_object',
	})
	.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```

await signAndExecuteTransactionBlock({ tx: txb });
```

### place

Places an item in the kiosk.

```typescript
const item = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

> /// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.

Reviewer: ```suggestion
/// Assume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
```

const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

kioskTx
	.place({
		item,
		itemType,
	})
	.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```

await signAndExecuteTransactionBlock({ tx: txb });
```

### list

Lists an item for sale (the item must be in the kiosk).

```typescript
const itemId = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

> /// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.

Reviewer: ```suggestion
// Assume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
```

const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

kioskTx
	.list({
		itemId,
		itemType,
		price: 100000n,
	})
	.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```

await signAndExecuteTransactionBlock({ tx: txb });
```

### placeAndList

> List an item for sale by first placing it in the kiosk (places the item & lists it for sale). It's a

Reviewer: ```suggestion
List an item for sale by first placing it in the kiosk (places the item and lists it for sale). It's a
```

> short hand for `place()` and `list()`

Reviewer: ```suggestion
short hand for `place()` and `list()`.
```

```typescript
const item = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

> /// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.

Reviewer: ```suggestion
// Assume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
```

const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

kioskTx
	.placeAndList({
		itemId,
		itemType,
		price: 100000n,
	})
	.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```

await signAndExecuteTransactionBlock({ tx: txb });
```

### delist

Removes the listing, keeping the item placed in the kiosk.

```typescript
const itemId = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

/// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

kioskTx
	.delist({
		itemId,
		itemType,
	})
	.finalize();

// sign and execute transaction block.
await signAndExecuteTransactionBlock({ tx: txb });
```

### withdraw

Withdraw (all or specific amount) from a kiosk.

`amount`: Can be empty, which will withdraw all the funds.

```typescript

> /// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.

Reviewer: ```suggestion
// Assume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
```

const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

kioskTx
	.withdraw({
		address: 'address_to_transfer_funds',
		amount: 100000n,
	})
	.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```

await signAndExecuteTransactionBlock({ tx: txb });
```

### borrowTx (callback)

Borrows an item from a kiosk. This function follows the `callback` approach, similar to the
ownerCap. The return of the item happens automatically after the execution of the callback.

```typescript
const itemId = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

/// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

kioskTx
	.borrowTx(
		{
			itemId,
			itemType,
		},
		(item) => {
			txb.moveCall({
				target: '0xMyGame::hero::level_up',
				arguments: [item],
			});
		},
	)
	.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```

await signAndExecuteTransactionBlock({ tx: txb });
```

### borrow / return

> Similar to `borrowTx`, borrows an item from the kiosk, but returns 2 transaction arguments: `item` &

Reviewer: ```suggestion
Similar to `borrowTx`, borrows an item from the kiosk, but returns two transaction arguments: `item` &
```

`promise`. You can use the `item` in your PTBs, but you must always call the `return()` function
with the `item` and the `Promise`.

```typescript
const itemId = '0xHeroAddress';
const itemType = '0x..::hero::Hero';

> /// asume `kioskClient` and `cap` are supplied to the function as explained in the previous section.

Reviewer: ```suggestion
// Assume `kioskClient` and `cap` are supplied to the function as explained in the previous section.
```

const txb = new TransactionBlock();
const kioskTx = new KioskTransaction({ transactionBlock: txb, kioskClient, cap });

const [item, promise] = kioskTx.borrow({
	itemId,
	itemType,
});

txb.moveCall({
	target: '0xMyGame::hero::level_up',
	arguments: [item],
});

kioskClient
	.return({
		itemType,
		item,
		promise,
	})
	.finalize();

> // sign and execute transaction block.

Reviewer: ```suggestion
// Sign and execute transaction block.
```
